core/src/main/scala/water/fvec/FrameOps.scala

```diff
@@ -0,0 +1,138 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package water.fvec
+
+import water.Futures
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * High-level DSL proving user-friendly operations
+ * on top of H2O Frame.
+ */
+trait FrameOps {
+  self: H2OFrame =>
+
+  /** Functional type to transform vectors. */
+  type VecTransformation = (String, Vec) => Vec
+
+  /** Functional type to select vectors. */
+  type VecSelector = (String, Vec) => Boolean
+
+  /** Create a sub-frame based on the list of column names.
+   *
+   * @param columnNames name of columns which will compose a new frame
+   * @return a new H2O Frame composed of selected vectors
+   */
+  def apply(columnNames: Array[String]): H2OFrame = new H2OFrame(subframe(columnNames))
+
+  /** Create a sub-frame based on the list of column names.
+   *
+   * @param columnNames name of columns which will compose a new frame
+   * @return a new H2O Frame composed of selected vectors
+   */
+  def apply(columnNames: Symbol*): H2OFrame = apply(columnNames.map(_.name).toArray)
+
+  def apply(transformation: VecTransformation, colNames: Symbol*): H2OFrame =
+    apply(transformation, colNames.map(_.name).toArray)
+
+  def apply(transformation: VecTransformation, colNames: Array[String]): H2OFrame = {
+    apply(transformation, (name, _) => colNames.contains(name))
+  }
+
+  def apply(transformation: VecTransformation, selector: VecSelector, removeVec: Boolean = true): H2OFrame = {
+    val vecs = self.vecs()
+    val names = self.names()
+    val futures = new Futures()
+    vecs.indices.filter(idx => {
+      val name = names(idx)
+      val vec = vecs(idx)
+      // Get only indexes of selected vectors
+      selector(name, vec)
+    }).map(idx => {
```

Comment:
Could we merge content of `map` and `foreach`, in order to avoid materialization of the output from map?

Reply:
Maybe replace the whole statement with one "for comprehension" statement

Reply:
Thanks, should be easier to read now :)